### PR TITLE
asus-hid parent/child model changes

### PR DESCRIPTION
**Pull request by superm1**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
@hughsie if I may ask some advice.  On real HW it seems that the child is marked unconnected, but this is obviously not the case.  How to handle that?

Log snippet:
```
06:42:31.641 FuDeviceList         waiting 10000ms for replug
06:42:31.774 FuEngine             auto-removing backend device N-KEY Device [62d5d8280031f607f1db058da959a97f6a8e6d90]
06:42:31.775 FuDeviceList         waiting 10000ms for N-KEY Device device removal
06:42:31.775 FuEngine             resetting update motd timeout
06:42:31.776 FuEngine             auto-removing backend device Microcontroller 0 [105a998eae01e451237dc37974beab9362f0066f]
06:42:31.785 FuDeviceList         waiting 10000ms for Microcontroller 0 device removal
06:42:31.785 FuEngine             resetting update motd timeout
06:42:31.786 FuEngine             auto-removing backend device Microcontroller 1 [c3c241d741fa6d1bac3980e2d79183c6211fc39e]
06:42:31.787 FuDeviceList         waiting 10000ms for Microcontroller 1 device removal
06:42:31.787 FuEngine             resetting update motd timeout
06:42:39.809 FuEngine             adding emulation-tag to /sys/devices/pci0000:00/0000:00:08.1/0000:09:00.3/usb1/1-3
06:42:39.810 FuDeviceList         found existing device 62d5d8280031f607f1db058da959a97f6a8e6d90, reusing item
06:42:39.810 FuDeviceList         not adding GUID 40e5a299-fe27-5e3b-a4a3-6649aeef4875 to device, use FU_DEVICE_PRIVATE_FLAG_ADD_COUNTERPART_GUIDS if required
06:42:39.810 FuDeviceList         copying child 105a998eae01e451237dc37974beab9362f0066f to new device
06:42:39.810 FuDevice             setting parent of Microcontroller 0 [105a998eae01e451237dc37974beab9362f0066f] to be ITE Upgrade Mode [62d5d8280031f607f1db058da959a97f6a8e6d90]
06:42:39.810 FuDeviceList         copying child c3c241d741fa6d1bac3980e2d79183c6211fc39e to new device
06:42:39.810 FuDevice             setting parent of Microcontroller 1 [c3c241d741fa6d1bac3980e2d79183c6211fc39e] to be ITE Upgrade Mode [62d5d8280031f607f1db058da959a97f6a8e6d90]
06:42:39.810 FuDeviceList         ::changed 62d5d8280031f607f1db058da959a97f6a8e6d90 [ITE Upgrade Mode]
06:42:39.810 FuDeviceList         62d5d8280031f607f1db058da959a97f6a8e6d90 old device came back, clearing flag
06:42:39.810 FuEngine             resetting update motd timeout
06:42:39.823 FuDeviceList         waited for replug
06:42:39.823 FuEngine             install phase now install
06:42:39.823 FuDeviceList         no replug or re-enumerate required
06:42:39.823 FuEngine             update -> FuAsusHidChildDevice:
  DeviceId:             105a998eae01e451237dc37974beab9362f0066f
  CompositeId:          62d5d8280031f607f1db058da959a97f6a8e6d90
  Name:                 Microcontroller 0
  Status:               device-write
  Percentage:           2
  Guid:                 f7765461-3ff4-52a0-a4d8-d782b1f885ab ← USB\VID_0B05&PID_1ABE&PART_RC71LS
  Plugin:               asus_hid
  Protocol:             com.asus.hid
  Flags:                internal|updatable-hidden|unsigned-payload
  Problems:             update-in-progress
  Vendor:               ASUSTek Computer, Inc.
  VendorId:             USB:0x0B05
  VendorId:             USB:0x048D
  Version:              311
  VersionFormat:        plain
  Created:              2024-11-14
  Modified:             2024-11-14
  UpdateError:          Device has been removed, An update is in progress
  PhysicalId:           1-3
  LogicalId:            RC71LS
  BackendId:            /sys/devices/pci0000:00/0000:00:08.1/0000:09:00.3/usb1/1-3
  RemoveDelay:          10000
  AcquiesceDelay:       2500
  Order:                -1
  PrivateFlags:         use-proxy-fallback,registered,unconnected
  Inhibit:              [unconnected] Device has been removed
  Inhibit:              [update-in-progress] An update is in progress
  Proxy:
    Guid[quirk]:        8460d283-a19f-50ff-9751-ae6322e93293 ← USB\TYPE_USB_DEVICE
    Guid[quirk]:        92fe1410-3e2c-5f69-a2db-74455b1135f1 ← USB\DRIVER_usb
    Guid[quirk]:        5f0738ae-5e3a-5e94-99fb-62d1b0fefcb1 ← USB\VID_0B05
    Guid[quirk]:        571229d5-caed-5cbc-91c1-5c216601a194 ← USB\CLASS_03
    Guid[quirk]:        8eeac4f3-d98d-57aa-a1fc-b44e8d3d8ba3 ← USB\CLASS_03&SUBCLASS_01
    Guid[quirk]:        c33b161d-a538-57df-b1c8-90817ada2387 ← USB\CLASS_03&SUBCLASS_01&PROT_01
    Guid[counterpart]:  af300991-5946-5d67-8075-083a90a7e78b
    PhysicalId:         1-3
    BackendId:          /sys/devices/pci0000:00/0000:00:08.1/0000:09:00.3/usb1/1-3
    Vid:                0xb05
    Pid:                0x1abe
    RemoveDelay:        10000
    AcquiesceDelay:     2500
    CustomFlags:        use-runtime-version
    GType:              FuAsusHidDevice
    FirmwareSizeMax:    262.1 kB
    Order:              0
    PossiblePlugin:     asus_hid
    PrivateFlags:       use-runtime-version,registered,unconnected
    Inhibit:            [unconnected] Device has been removed
    Events:

06:42:39.862 FuEngine             install phase now attach
06:42:39.862 FuEngine             failed to attach device after failed update: failed to attach using asus_hid: failed to subclass open: failed to open /dev/bus/usb/001/021: No such file or directory
06:42:39.862 FuEngine             install phase now cleanup
06:42:39.862 FuDeviceList         no replug or re-enumerate required
06:42:39.862 FuEngine             cleanup -> FuAsusHidChildDevice:
  DeviceId:             105a998eae01e451237dc37974beab9362f0066f
  CompositeId:          62d5d8280031f607f1db058da959a97f6a8e6d90
  Name:                 Microcontroller 0
  Status:               device-write
  Percentage:           2
  Guid:                 f7765461-3ff4-52a0-a4d8-d782b1f885ab ← USB\VID_0B05&PID_1ABE&PART_RC71LS
  Plugin:               asus_hid
  Protocol:             com.asus.hid
  Flags:                internal|updatable-hidden|unsigned-payload
  Vendor:               ASUSTek Computer, Inc.
  VendorId:             USB:0x0B05
  VendorId:             USB:0x048D
  Version:              311
  VersionFormat:        plain
  Created:              2024-11-14
  Modified:             2024-11-14
  UpdateState:          failed
  UpdateError:          Device has been removed
  PhysicalId:           1-3
  LogicalId:            RC71LS
  BackendId:            /sys/devices/pci0000:00/0000:00:08.1/0000:09:00.3/usb1/1-3
  RemoveDelay:          10000
  AcquiesceDelay:       2500
  Order:                -1
  PrivateFlags:         use-proxy-fallback,registered,unconnected
  Inhibit:              [unconnected] Device has been removed
  Proxy:
    Guid[quirk]:        8460d283-a19f-50ff-9751-ae6322e93293 ← USB\TYPE_USB_DEVICE
    Guid[quirk]:        92fe1410-3e2c-5f69-a2db-74455b1135f1 ← USB\DRIVER_usb
    Guid[quirk]:        5f0738ae-5e3a-5e94-99fb-62d1b0fefcb1 ← USB\VID_0B05
    Guid[quirk]:        571229d5-caed-5cbc-91c1-5c216601a194 ← USB\CLASS_03
    Guid[quirk]:        8eeac4f3-d98d-57aa-a1fc-b44e8d3d8ba3 ← USB\CLASS_03&SUBCLASS_01
    Guid[quirk]:        c33b161d-a538-57df-b1c8-90817ada2387 ← USB\CLASS_03&SUBCLASS_01&PROT_01
    Guid[counterpart]:  af300991-5946-5d67-8075-083a90a7e78b
    PhysicalId:         1-3
    BackendId:          /sys/devices/pci0000:00/0000:00:08.1/0000:09:00.3/usb1/1-3
    Vid:                0xb05
    Pid:                0x1abe
    RemoveDelay:        10000
    AcquiesceDelay:     2500
    CustomFlags:        use-runtime-version
    GType:              FuAsusHidDevice
    FirmwareSizeMax:    262.1 kB
    Order:              0
    PossiblePlugin:     asus_hid
    PrivateFlags:       use-runtime-version,registered,unconnected
    Inhibit:            [unconnected] Device has been removed
    Events:
```

**Comment by superm1** [comment]
But that might be a red herring, and the actual issue is the proxy isn't updated.
```
failed to write-firmware: failed to open device: failed to subclass open: failed to open /dev/bus/usb/001/023: No such file or directory
```

File: plugins/asus-hid/asus-hid.quirk [hunk 1]
```diff
@@ -21,9 +21,11 @@ Flags = use-runtime-version
 Plugin = asus_hid
 Flags = is-bootloader,can-verify-image
 FirmwareSizeMax = 0x40000
+GType = FuAsusHidDevice
```
Comment (hughsie) [inline plugins/asus-hid/asus-hid.quirk:24]: I think if you use fu_plugin_set_device_gtype_default() you might be able to drop this.

Comment (superm1) [reply plugins/asus-hid/asus-hid.quirk]: fixed, thanks!

File: plugins/asus-hid/asus-hid.quirk [hunk 2]
```diff
@@ -26,4 +24,4 @@ FirmwareSizeMax = 0x40000
 [USB\VID_048D&PID_89DC]
 Plugin = asus_hid
 Flags = is-bootloader,can-verify-image
-FirmwareSizeMax = 0x40000
+FirmwareSizeMax = 0x40000
```
Comment (hughsie) [inline plugins/asus-hid/asus-hid.quirk:27]: I think you lost the newline.

File: plugins/asus-hid/fu-asus-hid-device.c [hunk 3]
```diff
@@ -78,10 +78,17 @@ fu_asus_hid_device_init_seq(FuAsusHidDevice *self, GError **error)
...
-		fu_hid_device_set_interface(FU_HID_DEVICE(device), 0);
-	else
-		fu_hid_device_set_interface(FU_HID_DEVICE(device), 2);
+	FuAsusHidDevice *self = FU_ASUS_HID_DEVICE(device);
+
+	fu_hid_device_set_interface(FU_HID_DEVICE(device), 0);
+
+	for (guint i = 0; i < self->num_mcu; i++) {
+		g_autoptr(FuDevice) dev_tmp =
+		    fu_asus_hid_child_device_new(fu_device_get_context(device), i);
```
Comment (hughsie) [inline plugins/asus-hid/fu-asus-hid-device.c:87]: if you pass the `proxy` rather then the context the ctx gets set automatically.

File: plugins/asus-hid/fu-asus-hid-plugin.c [hunk 4]
```diff
@@ -28,6 +28,7 @@ fu_asus_hid_plugin_constructed(GObject *obj)
 	FuContext *ctx = fu_plugin_get_context(plugin);
 	fu_plugin_add_device_gtype(plugin, FU_TYPE_ASUS_HID_DEVICE);
 	fu_plugin_add_device_gtype(plugin, FU_TYPE_ASUS_HID_CHILD_DEVICE); /* coverage */
+	fu_plugin_set_device_gtype_default(plugin, FU_TYPE_ASUS_HID_DEVICE);
```
Comment (hughsie) [inline plugins/asus-hid/fu-asus-hid-plugin.c:31]: if you do this, you don't need `fu_plugin_add_device_gtype(plugin, FU_TYPE_ASUS_HID_DEVICE);`

File: plugins/asus-hid/fu-asus-hid-device.c (hunk 3 above)
Comment (superm1) [reply plugins/asus-hid/fu-asus-hid-device.c]: ack, switched

File: plugins/asus-hid/fu-asus-hid-child-device.c [hunk 5]
```diff
@@ -276,9 +281,14 @@ fu_asus_hid_child_device_class_init(FuAsusHidChildDeviceClass *klass)
 }
 
 FuDevice *
-fu_asus_hid_child_device_new(FuContext *ctx, guint8 idx)
+fu_asus_hid_child_device_new(FuDevice *proxy, guint8 idx)
 {
-	FuDevice *dev = g_object_new(FU_TYPE_ASUS_HID_CHILD_DEVICE, "context", ctx, NULL);
+	FuDevice *dev = g_object_new(FU_TYPE_ASUS_HID_CHILD_DEVICE,
+				     "context",
+				     fu_device_get_context(proxy),
```
Comment (hughsie) [inline plugins/asus-hid/fu-asus-hid-child-device.c:288]: I don't think you need these two lines

File: plugins/asus-hid/fu-asus-hid-device.c [hunk 6]
```diff
@@ -78,10 +78,15 @@ fu_asus_hid_device_init_seq(FuAsusHidDevice *self, GError **error)
...
-		fu_hid_device_set_interface(FU_HID_DEVICE(device), 0);
-	else
-		fu_hid_device_set_interface(FU_HID_DEVICE(device), 2);
+	FuAsusHidDevice *self = FU_ASUS_HID_DEVICE(device);
+
+	fu_hid_device_set_interface(FU_HID_DEVICE(device), 0);
+
+	for (guint i = 0; i < self->num_mcu; i++) {
+		g_autoptr(FuDevice) dev_tmp = fu_asus_hid_child_device_new(device, i);
+
```
Comment (hughsie) [inline plugins/asus-hid/fu-asus-hid-device.c:87]: maybe nuke the empty line?

**Review by hughsie — DISMISSED** [review]
two nits, then LGTM, thanks.

**Comment by superm1** [comment]
> two nits, then LGTM, thanks.

thx.  any ideas on who should be updating the proxy after detach?  I think that's what's blocking on real hardware at the moment (emulation works E2E on one controller with this branch)

**Comment by hughsie** [comment]
@superm1 sorry I missed your question above; one thing that does look fishy is the calls to: `	FuDevice *parent = fu_device_get_parent(device);` --- that should be the proxy, no?

I think I understand what's going on; the AsusHidNumMcu is set in the runtime (so the children get created) but not in the bootloader -- so we use the children from the runtime when in bootloader mode. That's not crazy, but I do think we're missing something in src/fu-device-list.c if we want it to be magic. I'm hesitant as it might break something that relies on the proxy being different to the parent.

The slightly more safe suggestion would be to watch the "FuDevice::child-added" signal in the bootloader device and then in that callback set the proxy for the child there.

**Pull request merged**